### PR TITLE
WIP:Add OSGi support to OpenTelemetry project

### PR DESCRIPTION
**Pull request by LWogan**
Use bnd plugin to add OSGi metdata to manifest files to allow them to be used within an OSGi framework

### CODE REVIEW

**Comment by jkwatson** [comment]
@LWogan before we can look at this or accept it, we need you to sign the CLA. Thanks!

**Comment by jkwatson** [comment]
I know virtually nothing about OSGi, but this looks ok to me on its surface. @jack-berg do you have any experience with OSGi, or know anyone we can rope in to help?

**Comment by LWogan** [comment]
To add a little explanation of what did does. The bnd instruction will generate OSGi metadata and add it to the manifest files of each jar.  OSGi is a framework which relies on this metadata to functionally operate. The OpenTelemetry artifacts cannot be used by anyone running in an OSGi framework without this. 

There was some logic already present in the project to set specific manifest attributes, I have preserved these by instructing the bnd operation to add them while adding its automatically generated OSGi metadata.

File: buildSrc/src/main/kotlin/otel.java-conventions.gradle.kts [hunk 1]
```diff
@@ -126,6 +124,15 @@ tasks {
     }
   }
 
+  named("jar", Jar::class.java) {
```
Comment (jack-berg) [inline buildSrc/src/main/kotlin/otel.java-conventions.gradle.kts:127]: Can you add a comment indicating that this adds required OSGi metadata to the jar manifest?

**Review by jack-berg — APPROVED** [review]
I'm not familiar with OSGi but I asked around New Relic and a couple peek took a glance and said it looks innocent. 

For reference, here's what the `:api` jar manifest looks like before:
```
Manifest-Version: 1.0
Automatic-Module-Name: io.opentelemetry.api
Built-By: jberg
Built-JDK: 11.0.15
Implementation-Title: all
Implementation-Version: 1.17.0-SNAPSHOT
```

and after:
```
Manifest-Version: 1.0
Automatic-Module-Name: io.opentelemetry.api
Bnd-LastModified: 1658958870198
Built-By: jberg
Built-JDK: 11.0.15
Bundle-ManifestVersion: 2
Bundle-Name: opentelemetry-api
Bundle-SymbolicName: opentelemetry-api
Bundle-Version: 1.17.0.SNAPSHOT
Created-By: 11.0.15 (Eclipse Adoptium)
Implementation-Title: all
Implementation-Version: 1.17.0-SNAPSHOT
Import-Package: io.opentelemetry.context,io.opentelemetry.context.prop
 agation,io.opentelemetry.sdk.autoconfigure,javax.annotation;version="
 [3.0,4)"
Private-Package: io.opentelemetry.api,io.opentelemetry.api.baggage,io.
 opentelemetry.api.baggage.propagation,io.opentelemetry.api.common,io.
 opentelemetry.api.internal,io.opentelemetry.api.metrics,io.openteleme
 try.api.trace,io.opentelemetry.api.trace.propagation,io.opentelemetry
 .api.trace.propagation.internal
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Tool: Bnd-6.3.1.202206071316
```

**Comment by LWogan** [comment]
@jack-berg Ive made some additional changes. I needed to add annotations to numerous package-info files. This informs the OSGi framework that these packages are available for use by other jars. Worth noting some packages marked with the namespace internal have been included as some of their classes are on the APIs of classes in other packages. 

Another point to add is this PR will add support to this project for OSGi. However there are numerous dependencies such as io.grpc (and its transitive dependencies) which also do not have OSGi support. There are workarounds to this but I thought I should point that out.

Leaving as a draft for now as I am still integrating OpenTelemetry into an OSGi project so I might not have worked out all the issues yet.

**Comment by splatch** [comment]
I was able to test it and I believe it needs a second touch. Import of `javax.annotation` package should be marked as optional. Otherwise runtime environment is forced to host findbugs which is strictly speaking build time dependency/tool which does not serve any purpose for execution of byte code.

**Comment by splatch** [comment]
@jack-berg Looking at this again (on top of 0.16) I think there are more things to tackle within library design itself.

The `SpanExporterConfiguration` introduces a lookup of a specific implementations from a dependency consumed in opentelemetry-api module:
- [SpanExporterConfiguration.configureSpanExporters](https://github.com/open-telemetry/opentelemetry-java/blob/v1.16.0/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/SpanExporterConfiguration.java#L88L107)
- [SpanExporterConfiguration.configureExporter](https://github.com/open-telemetry/opentelemetry-java/blob/v1.16.0/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/SpanExporterConfiguration.java#L112L157)
- [SpanExporterConfiguration.configureOtlp](https://github.com/open-telemetry/opentelemetry-java/blob/v1.16.0/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/SpanExporterConfiguration.java#L162L180)
- [SpanExporterConfiguration.configureZipkin](https://github.com/open-telemetry/opentelemetry-java/blob/v1.16.0/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/SpanExporterConfiguration.java#L184L200)

Its not quite clear to me why such design exists. With such way in place you tie autconfigure module with all listed spi providers. Effectively it is not possible to separate api from specific implementations as bndtools which generates OSGi metadata is able to spot somehow `Class.forName` and refered class names. Effectively it causes direct listing of such packages as import statements for for sdk-extensions-autoconfigure, which is used by api. Wouldn't it be better to have some kind of `ExporterFactory` looked up through `ServiceLoader` first and then let it configure specific exporter through passed properties?

Any use of `Class.forName(String)` without a `ClassLoader` is a trouble marker under OSGi cause it will fail in most of the cases:
- [ClasspathUtil.checkClassExists](https://github.com/open-telemetry/opentelemetry-java/blob/v1.16.0/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ClasspathUtil.java)
- [GlobalOpenTelemetry.maybeAutoConfigureAndSetGlobal](https://github.com/open-telemetry/opentelemetry-java/blob/v1.16.0/api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java#L217)

Such patterns work fine with single classloader which is used with flat classpath, but will certainly fail with hierarchical classloaders and osgi which has multiple classloaders. For simplicity of understanding - under OSGi each bundle (deployed jar) have its own classloader exposing only declared import statements.

**Comment by jack-berg** [comment]
> Its not quite clear to me why such design exists. 

Hi @splatch - sorry for the delayed reply. Not sure what the best answer is here, but I can provide some context. The [autoconfigure](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure) is responsible for configuring an `OpenTelemetrySdk` instance based on system properties and env vars. The components it configures includes exporters for specific protocols, such as zipkin, jaeger, prometheus, otlp, etc. Autoconfigure doesn't want to impose transitive dependencies on those exporters, since users will likely only need one or two and the others would just result in unnecessary bulk. So instead it places `compileOnly` dependencies on these components, and executes the configuration logic for each conditionally based on its presence on the classpath. 

The advantage of this is that configuration lives in one place, instead of being scattered around each of the components. Not sure how you could accomplish this with the `ExporterFactory` pattern you describe, but admittedly I may not be fully grasping your suggestion.

**Comment by PascalDutch** [comment]
Hi, why is this never merged? We are also looking for functionality like this, since we want to use OpenTelemetry in an OSGi context.

**Comment by LWogan** [comment]
> Hi, why is this never merged? We are also looking for functionality like this, since we want to use OpenTelemetry in an OSGi context.

From my perspective the pr is incomplete. Needs more testing and probably more changes. 

The package structures of this Otel project means if you want to pull in a single lib then you pretty much have to pull in every other subproject due to the internal dependency tree with each subproject having some package import from another. This is a frustration when you only wish to use a subset of bundles. 

Also, unfortunately, many of OpenTelems dependencies and also transient dependencies are not OSGi bundles which makes it very difficult to use this library within an OSGi context even with this PR applied.

**Pull request closed without merging**